packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,229 @@
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const vub_suggestion =
+        tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = vub_suggestion;
+        return {
+          valid: true,
+          validaUntilBlock: {
+            fixed: true,
+            prev: validUntilBlock,
+            suggestion: vub_suggestion
+          }
+        };
+      }
+      return {
+        valid: false,
+        validaUntilBlock: {
+          fixed: false,
+          prev: validUntilBlock,
+          suggestion: vub_suggestion
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  };
+
+  // validateIntents = async (): Promise<ValidationResult> => {
+  //   return {
+  //     result: true,
+  //     fixed: false
+  //   };
+  // };
+
+  /**
+   * validate systemFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  validateSystemFee = async (autoFix = false): Promise<ValidationResult> => {
+    const { script, systemFee } = this.transaction;
+    const { gas_consumed } = await this.rpcClient.invokeScript(script);
+    const minimumSystemFee = new u.Fixed8(parseFloat(gas_consumed)).ceil();
+    if (autoFix && !minimumSystemFee.equals(systemFee)) {
+      this.transaction.systemFee = minimumSystemFee;
+      return {
+        valid: true,
+        systemFee: {
+          fixed: true,
+          prev: systemFee,
+          suggestion: minimumSystemFee
+        }
+      };
+    } else if (minimumSystemFee.isGreaterThan(systemFee)) {
+      return {
+        valid: false,
+        systemFee: {
+          fixed: false,
+          prev: systemFee,
+          suggestion: minimumSystemFee
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  };
+
+  /**
+   * Validate NetworkFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  validateNetworkFee = async (autoFix = false): Promise<ValidationResult> => {
+    const { networkFee } = this.transaction;
+    const minimumNetworkFee = getNetworkFee(this.transaction);
+    if (autoFix && !minimumNetworkFee.equals(networkFee)) {
+      this.transaction.networkFee = minimumNetworkFee;
+      return {
+        valid: true,
+        networkFee: {
+          fixed: true,
+          prev: networkFee,
+          suggestion: minimumNetworkFee
+        }
+      };
+    } else if (minimumNetworkFee.isGreaterThan(networkFee)) {
+      return {
+        valid: false,
+        networkFee: {
+          fixed: false,
+          prev: networkFee,
+          suggestion: minimumNetworkFee
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  };
+
+  /**
+   * Validate Signatures
+   */
+  validateSigning = async (): Promise<ValidationResult> => {
```

Comment:
Signatures are part of transaction and signing should be done before validation I think.

Reply:
I thought we have discussed this multiple times. The order should be build, validate and sign. Signing must always be the last step because any adjustments during the build and validate step will require a resigning.

Reply:
I have removed validateSigning, however how can we add such empty witnesses?